source/fe/fe_hermite.cc

```diff
@@ -0,0 +1,840 @@
+      const typename Mapping<2, spacedim>::InternalDataBase &mapping_data,
+      Table<2, Number> &                                     value_list)
+    {
+      unsigned int n_q_points;
+      Tensor<1, 2> cell_extents;
+
+      // Check mapping_data is associated with a compatible mapping class
+      if (dynamic_cast<const typename MappingHermite<2, spacedim>::InternalData
+                         *>(&mapping_data) != nullptr)
+        {
+          const typename MappingHermite<2, spacedim>::InternalData *data =
+            dynamic_cast<
+              const typename MappingHermite<2, spacedim>::InternalData *>(
+              &mapping_data);
+          n_q_points   = data->quadrature_points.size();
+          cell_extents = data->cell_extents;
+        }
+      else if (dynamic_cast<const typename MappingCartesian<2>::InternalData *>(
+                 &mapping_data) != nullptr)
+        {
+          const typename MappingCartesian<2>::InternalData *data =
+            dynamic_cast<const typename MappingCartesian<2>::InternalData *>(
+              &mapping_data);
+          n_q_points   = data->quadrature_points.size();
+          cell_extents = data->cell_extents;
+        }
+      else
+        Assert(false, ExcInternalError());
+
+      const unsigned int dofs         = fe_herm.n_dofs_per_cell();
+      const unsigned int regularity   = fe_herm.get_regularity();
+      const unsigned int dofs_per_dim = 2 * regularity + 2;
+      const unsigned int q_points     = value_list.size(1);
+      (void)dofs;
+
+      AssertDimension(value_list.size(0), dofs);
+      AssertDimension(dofs_per_dim * dofs_per_dim, dofs);
+      AssertIndexRange(q_points, n_q_points + 1);
+      (void)n_q_points;
+
+      std::vector<unsigned int> l2h =
+        hermite_lexicographic_to_hierarchic_numbering<2>(regularity);
+
+      AssertDimension(l2h.size(), dofs);
+
+      FullMatrix<double> factors(dofs_per_dim, dofs_per_dim);
+
+      for (unsigned int q = 0; q < q_points; ++q)
+        {
+          double factor_2 = 1.0;
+
+          for (unsigned int d3 = 0, d4 = regularity + 1; d4 < dofs_per_dim;
+               ++d3, ++d4)
+            {
+              double factor_1 = factor_2;
+
+              for (unsigned int d1 = 0, d2 = regularity + 1; d2 < dofs_per_dim;
+                   ++d1, ++d2)
+                {
+                  factors(d1, d3) = factor_1;
+                  factors(d2, d3) = factor_1;
+                  factors(d1, d4) = factor_1;
+                  factors(d2, d4) = factor_1;
+
+                  value_list(l2h[d1 + d3 * dofs_per_dim], q) *= factor_1;
+                  value_list(l2h[d2 + d3 * dofs_per_dim], q) *= factor_1;
+                  value_list(l2h[d1 + d4 * dofs_per_dim], q) *= factor_1;
+                  value_list(l2h[d2 + d4 * dofs_per_dim], q) *= factor_1;
+
+                  factor_1 *= cell_extents[0];
+                }
+
+              factor_2 *= cell_extents[1];
+            }
+        }

```

Comment:
I know this type of code^^ Could you add some comments that you have unrolled the loop over nodes.